### PR TITLE
Rework SavedView data structure to enable lazy data loading

### PR DESCRIPTION
**Pull request by roluk**
Thumbnails and Saved View data is now stored outside `SavedView` object. This allows us to supply Saved View metadata to consumers early and load the rest on demand.

`useSavedViews` now produces placeholder thumbnail `ReactNode` objects that will load the image and view data once the thumbnail area appears on screen.

### Beaking changes

* `SavedView` interface changes
  * Rename `id` property to `savedViewId`
  * Remove `thumbnail` property
  * Move `viewData` and `extensions` properties to `SavedViewData` type
    * Type of `viewData` has been changed and definition has moved to `@itwin/saved-views-react`
  * Change type of `creationTime` and `lastModified` properties to `Date | undefined`
* `useSavedViews` hook rework
  * No longer implements optimistic behaviour
  * Lazily loads Saved View thumbnails and `SavedViewData`
  * Can utilize user-supplied external state store
  * All actions now return a promise that resolves on action completion
  * All creation actions now return id of the created entity
  * Rename `SavedViewActions` type to `SavedViewsActions`
  * Split `submitSavedView` into `createSavedView` and `updateSavedView`
  * Add `lookupSavedViewData` action
  * Remove `moveToNewGroup` and `addNewTag` actions
* `SavedViewTag`: Rename `id` property to `tagId`
* `SavedViewGroup`: Rename `id` property to `groupId`
* `SavedViewsClient` interface changes
  * Rename methods
    * `getAllSavedViews` -> `getSavedViews`
    * `getAllGroups` -> `getGroups`
    * `getAllTags` -> `getTags`
    * `getSavedView` -> `getSavedViewById`
  * Add `getSavedViewDataById` method
  * Method argument types now instead of ending in `*Params` now end in `*Args`, e.g. `CreateSavedViewParams` -> `CreateSavedViewArgs`
  * You can now query all Saved Views that are assigned to a particular group using `SavedViewsClient.getSavedViews`
  * `CreateSavedViewArgs`, `UpdateSavedViewArgs`: Inline properties previously nested within `savedView` property
  * `CreateGroupArgs`, `UpdateGroupArgs`: Inline properties previously nested within `group` property
  * `UpdateTagArgs`: Inline properties previously nested within `tag` property
* `<SavedViewTile />` changes
  * Add `thumbnail` prop
  * Thumbnails that were specified as image URLs now need to be explicitly passed as ``
  * `onRename` callback will no longer send `undefined` for `newName` argument

### Other changes

* Update TypeScript ^5.3.3 -> ^5.5.4
* Update ESLint ^8.45 -> ^8.57
* Remove redundant browser history entry when navigating to components catalog

### CODE REVIEW

File: packages/saved-views-react/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,43 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
...
 ## [Unreleased](https://github.com/iTwin/saved-views/tree/HEAD/packages/saved-views-react)
 
+### Beaking changes
+
+* `SavedView` interface changes
+  * Rename `id` property to `savedViewId`
+  * Remove `thumbnail` property
+  * Move `viewData` and `extensions` properties `SavedViewData` type
+    * Type of `viewData` has been changed and definition has moved to `@itwin/saved-views-react`
+  * Change type of `creationTime` and `lastModifier` properties to `Date | undefined`
```
Comment (diegopinate) [inline packages/saved-views-react/CHANGELOG.md:16]:
```suggestion
  * Change type of `creationTime` and `lastModified` properties to `Date | undefined`
```

Comment (roluk) [reply packages/saved-views-react/CHANGELOG.md]: Fixed in b892f140abfcddf4289c5543f72b4bbbc9babfd9.

File: packages/saved-views-react/src/SavedViewsClient/ITwinSavedViewsClient.ts [hunk 2]
```diff
@@ -69,94 +71,104 @@ export class ITwinSavedViewsClient implements SavedViewsClient {
...
       signal: args.signal,
     });
 
-    await Promise.all((args.savedView.extensions ?? []).map(
-      ({ extensionName, data }) => this.#client.createExtension({
-        savedViewId: args.savedView.id,
-        extensionName,
-        data,
-      }),
+    await Promise.all((args.savedViewData?.extensions ?? []).map(
```
[the remark this answers is not in the document]
Comment (roluk) [reply packages/saved-views-react/src/SavedViewsClient/ITwinSavedViewsClient.ts]: Added line breaks in 9fb987c789e406dd63b18e543458094fe5cc086a.

File: packages/saved-views-react/src/SavedView.ts [hunk 3]
```diff
@@ -2,23 +2,36 @@
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
-import type { ViewData } from "@itwin/saved-views-client";
-import type { ReactNode } from "react";
+import type { ViewITwin3d, ViewITwinDrawing, ViewITwinSheet } from "@itwin/saved-views-client";
 
 export interface SavedView {
```
[the remark this answers is not in the document]
Comment (roluk) [reply packages/saved-views-react/src/SavedView.ts:7]: There are many arguments for and against using interfaces versus types. In the end I don't ever care which one is used.

Here however, I think the fact that user can extend our types is a nice feature to have.

File: packages/saved-views-react/CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,43 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased](https://github.com/iTwin/saved-views/tree/HEAD/packages/saved-views-react)
 
+### Beaking changes
+
+* `SavedView` interface changes
+  * Rename `id` property to `savedViewId`
+  * Remove `thumbnail` property
+  * Move `viewData` and `extensions` properties `SavedViewData` type
```
[the remark this answers is not in the document]
Comment (roluk) [reply packages/saved-views-react/CHANGELOG.md]: Fixed in b892f140abfcddf4289c5543f72b4bbbc9babfd9.

File: packages/test-app-frontend/src/App/ComponentsCatalog/useSavedViewData.tsx [hunk 5]
```diff
@@ -250,15 +171,26 @@ export function useThumbnailImage(): string | undefined {
...
 }
 
-async function getThumbnailImage(): Promise<string | undefined> {
+async function getThumbnailImage(): Promise<Blob> {
   const canvas = new OffscreenCanvas(400, 300);
 
   const ctx = canvas.getContext("2d");
   if (!ctx) {
-    return undefined;
+    throw new Error("Could not obtain 2d context from offscreen canvas.");
```
[the remark this answers is not in the document]
Comment (roluk) [reply packages/test-app-frontend/src/App/ComponentsCatalog/useSavedViewData.tsx:193]: This is component playground side of the application. It's a rarity for a computer or browser not to support 2d canvas context and this error serves more as a troubleshooting help if anything.

File: packages/saved-views-react/src/SavedViewsClient/ITwinSavedViewsClient.ts [hunk 6]
```diff
@@ -166,39 +182,36 @@ export class ITwinSavedViewsClient implements SavedViewsClient {
...
   }
 
-  async deleteTag(args: DeleteTagParams): Promise<void> {
+  async deleteTag(args: DeleteTagArgs): Promise<void> {
     await this.#client.deleteTag({ tagId: args.tagId, signal: args.signal });
   }
 }
 
-function savedViewResponseToSavedView(response: SavedViewRepresentation): SavedView {
+function savedViewResponseToSavedView(response: Omit<SavedViewMinimal, "savedViewData" | "extensions">): SavedView {
```
[the remark this answers is not in the document]
Comment (roluk) [reply packages/saved-views-react/src/SavedViewsClient/ITwinSavedViewsClient.ts:199]: [Our API documentation](https://developer.bentley.com/apis/savedviews/operations/get-savedview/) claims that a minimal Saved View representation will have full `savedViewData` information, but only reference links for `extensions`. I verified this behaviour in QA.

**Review by diegopinate — APPROVED** [review]

**Pull request merged**